Foundations for the budget automations UI

jfdoming opened the pull request:
Round 2 of prototyping a goal template UI (the [previous PR](https://github.com/actualbudget/actual/pull/3918) can't be reopened sadly). Addressed a bit of feedback from that PR.

youngcw commented:
Looks good so far.  One thing that maybe you could keep in mind is that it would be nice to have a central place to view and edit templates.  So this new modal to make templates should be ready to be used both from the budget page and a goals list page.

youngcw commented:
Do you have thoughts on gracefully migrating from the notes to the UI?  Would it end up being a all at once thing once the UI is in a good state?

jfdoming commented:
> Looks good so far. One thing that maybe you could keep in mind is that it would be nice to have a central place to view and edit templates. So this new modal to make templates should be ready to be used both from the budget page and a goals list page.

Curious, what is the intended use-cases for an overall goals list? I'm not necessarily opposed, but I do worry that it's yet-another-page and that it would have a lot of visual noise

jfdoming commented:
> Do you have thoughts on gracefully migrating from the notes to the UI? Would it end up being a all at once thing once the UI is in a good state?

Good question! Here was my overall framing for the rollout:
- With goal templates being so widely used already, we should decouple the rollout of UI features from the release/non-release of regular templates. (E.g., how in this PR the UI is an experimental sub-feature)
- The above gives us freedom to incrementally add features to this UI, i.e., not build all template types/options at once.
- In order to take advantage of that freedom, we'd need to let people _partially_ migrate so they don't lose functionality by switching to the UI.

My thinking for partial migrations is to show a pop-up when you open the UI in a category that currently has note-based templates, asking the user if they'd like to switch that category to the UI instead. The downside of course is it's a bit toilsome to do for all categories—maybe that pop-up can have a "do for all" checkbox or something. Wdyt?

youngcw commented:
> > Looks good so far. One thing that maybe you could keep in mind is that it would be nice to have a central place to view and edit templates. So this new modal to make templates should be ready to be used both from the budget page and a goals list page.
> 
> Curious, what is the intended use-cases for an overall goals list? I'm not necessarily opposed, but I do worry that it's yet-another-page and that it would have a lot of visual noise

The idea would be that it's a way to plan out your budget. You could see the total expected budget amount, see the order of the templates if you have many priority levels, add/adjust templates while being able to see the big picture, etc.

Teprifer commented:
> > Looks good so far. One thing that maybe you could keep in mind is that it would be nice to have a central place to view and edit templates. So this new modal to make templates should be ready to be used both from the budget page and a goals list page.
> 
> Curious, what is the intended use-cases for an overall goals list? I'm not necessarily opposed, but I do worry that it's yet-another-page and that it would have a lot of visual noise

It'd make it a lot easier to see at a glance, and make priorities significantly easier to plan out - either by being able to sort by priority, or drag and drop. See totals per category, per group, budget overall, that sort of budget nerdy stuff. A separate feature to this PR.

---

Regarding migrating, how about a multi-stage approach?

Initially as the UI is still being built, rather than ask if the user wants to switch the category over to the UI, which would be annoying indeed, instead have a warning at the top of the UI advising that note templates are disabled for the category while UI automations are in use. You'd have to tweak the UI slightly to not start with an automation box. 
This removes asking, and still allows a category by category approach to migrating.

Once the UI can handle all templates the check and apply template menus pops up a notification that text based templates are still in use and we recommend converting them to UI templates as text based will be removed in the future.
Could do things like colour the notes icon red or something to indicate which need converting. 

Finally, several months after the UI is complete, remove the text based templates so there's less to maintain.

---

Suggest a ? or help link within the UI to explain what budget automations are, or link to the docs page.

ftbboy2115 commented:
I would love to see the Templates/Goals feature moved from Experimental to Main. Keep up the great work! 
I haven't messed with the experimental features, but read about them in the docs. I like what YNAB is doing with them and would love to see them in Actual 😊

jfdoming commented:
@youngcw @Teprifer finally got around to addressing the previous comments.

For a rollout plan, here's what I have in mind—let me know your thoughts:
1. We merge this and some other UI PRs behind a "dev-only" feature flag, which I've implemented by requiring you to set `devEnableGoalTemplatesUI` in `localStorage` to the value `"true"` (i.e., `localStorage.setItem("devEnableGoalTemplatesUI", "true")`). This is to get some changes out the door as there is a lot of UI to build and things to consider carefully that I don't want to cram into a small series of PRs.
2. We implement the ability to migrate a single category to/from UI-based automations (with preferences for each category); maybe we can cram a set of buttons into the modal to start. Thinking to take Teprifer's idea to colour the notes icon red if UI is active for a category and yet it contains template text.
3. Eventually, we implement support for all automation types and move the feature from "dev" to a standard feature flag only.
4. Much later, after extensive testing, we create a release with a note calling out _removal_ of support for notes-based templates in the subsequent release, and (looking for feedback if this is needed) add a new advanced setting to "Migrate all notes-based templates" and remove the previous migration buttons and support for _running_ text-based templates (but still detect them in red).
5. After the above, we release goal templates and the automation UI.
6. Eventually, we remove support for detecting templates along with all the parsing logic/etc.

Assuming the above plan is ok with y'all, I've moved most changes to a separate PR https://github.com/actualbudget/actual/pull/4398 and will un-mark this PR as draft! Looking to see if the overall approach addresses your comments (note that I haven't implemented priorities or visualizing templates yet, but the structure is there—the idea for the latter is you'll have some way of temporarily overriding your "Available funds" value for the month and can visualize goals from there). Preview link [here](https://deploy-preview-4398.demo.actualbudget.org/)

jfdoming commented on packages/desktop-client/src/components/budget/goals/CategoryAutomationButton.tsx:
```diff
@@ -0,0 +1,57 @@
+  defaultColor = theme.buttonNormalText,
+  style,
+}: CategoryAutomationButtonProps) {
+  const automations: Template[] = [];
+  const hasAutomations = !!automations.length;
```
This is intentional while we're building this feature. It will be changed in a future PR.

jfdoming commented on packages/loot-core/src/server/budget/types/templates.d.ts:
```diff
@@ -60,7 +60,7 @@ interface AverageTemplate extends BaseTemplate {
 
 interface GoalTemplate extends BaseTemplate {
-  type: 'simple';
+  type: 'goal';
```
Fixed since I'm pretty sure it's a typo, will matter later once we implement goal-type templates

jfdoming commented on packages/loot-core/src/types/prefs.d.ts:
```diff
@@ -1,5 +1,6 @@
 export type FeatureFlag =
   | 'goalTemplatesEnabled'
+  | 'goalTemplatesUIEnabled'
```
Calling this `goalTemplatesUIEnabled` rather than `budgetAutomationsUIEnabled` so the flags are consistent (and a search in the codebase for `goalTemplates` in the future will find all usages)

jfdoming commented on packages/desktop-client/src/components/budget/goals/CategoryAutomationButton.tsx:
```diff
@@ -0,0 +1,57 @@
+  defaultColor?: string;
+  tooltipPosition?: ComponentProps<typeof Tooltip>['placement'];
+  style?: CSSProperties;
+};
```
Nice catch

youngcw commented:
That process seems good to me.  Could we have the feature be available when in development mode instead of having to set a variable every time we want to test? Similar to the theme toggle button that is only visible in dev builds.

jfdoming commented:
> That process seems good to me. Could we have the feature be available when in development mode instead of having to set a variable every time we want to test? Similar to the theme toggle button that is only visible in dev builds.

Good idea

youngcw commented:
I think the only issue with that approach is that the demo builds won't work that way and would still need the dev variable via the console or something.

jfdoming commented:
> I think the only issue with that approach is that the demo builds won't work that way and would still need the dev variable via the console or something.

One other thought is that if we do dev mode then anyone with a local install will see the option, no? Maybe we need to just do it on preview URLs :/

youngcw commented:
I guess the manual setting works for now.  We can change it later to something else if we want.

matt-fidd commented:
Would you consider keeping the text functionality too in the background?

Moving it out of notes would be good, but maybe the automations modal could have a toggle between editing it in graphical/text format with graphical as the default of course.

youngcw commented:
It could be nice to be able to add via text in the UI, at least at first.  I would like to ultimately move away from the text templates as the text parsing limits potential flexibility of the templates.

matt-fidd commented:
> It could be nice to be able to add via text in the UI, at least at first. I would like to ultimately move away from the text templates as the text parsing limits potential flexibility of the templates.

That's fair enough, if it becomes limiting we should drop them.

youngcw commented:
One example of text being limiting is that the schedule templates need extra special characters to parse the new percent adjustments properly since the schedule name can be anything, making it hard to parse with extra flags.

matt-fidd commented:
I noticed that, I didn't spot the PR in time but I was going to suggest dropping those and quoting the schedule name instead (if that's possible)

#template schedule "Long Schedule name with Percent in the title" increase 10%

jfdoming commented:
One other observation: the UI I'm proposing already has a textual representation of the templates (templates are summarized when outside of the editor). So maybe we can rely on that instead?

ftbboy2115 commented:
> It could be nice to be able to add via text in the UI, at least at first. I would like to ultimately move away from the text templates as the text parsing limits potential flexibility of the templates.

I can imagine a lot of work would be needed to be able to parse the text to pick out flags for templates, while using that data field for other uses as well. But I also really like having the ability to add notes to each goal/category, group, transaction, etc. Would we be able to keep the note-taking ability while adding a separate data point for the goal?

jfdoming commented:
> > It could be nice to be able to add via text in the UI, at least at first. I would like to ultimately move away from the text templates as the text parsing limits potential flexibility of the templates.
> 
> I can imagine a lot of work would be needed to be able to parse the text to pick out flags for templates, while using that data field for other uses as well. But I also really like having the ability to add notes to each goal/category, group, transaction, etc. Would we be able to keep the note-taking ability while adding a separate data point for the goal?

The parsing is how templates work today. This thread is planning on eventually _removing_ text-based templates. There is no plan to remove notes.
